### PR TITLE
Improve halo_n and u_prime

### PR DESCRIPTION
**Pull request by wborgeaud**
From this discussion: https://github.com/mir-protocol/plonky/pull/41#discussion_r442426267
Fixed of the definition of `u_prime` to include scaling with endo. I also wrote the implementation of the paper's Algorithm 1 out-of-circuit.

### CODE REVIEW

File: src/curve.rs [hunk 1]
```diff
@@ -44,6 +44,18 @@ pub trait Curve: 'static + Sync + Sized + Copy + Debug {
 pub trait HaloCurve: Curve {
     const ZETA: Self::BaseField;
     const ZETA_SCALAR: Self::ScalarField;
+
+    fn endomorphism(p: AffinePoint<Self>) -> AffinePoint<Self> {
```
Comment (dlubarov) [inline src/curve.rs:48]: Maybe have it take `&self`, or do you think `p.endomorphism()` doesn't look as natural?

File: src/field.rs [hunk 2]
```diff
@@ -130,7 +130,12 @@ pub trait Field:
     }
 
     fn from_canonical_bool(b: bool) -> Self {
-        Self::from_canonical_u64(if b { 1 } else { 0 })
+        if b {
+            Self::ONE
+        } else {
+            Self::ZERO
+        }
+        // Self::from_canonical_u64(if b { 1 } else { 0 })
```
Comment (dlubarov) [inline src/field.rs:138]: I like your version; we can delete this

File: src/plonk_util.rs [hunk 3]
```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
     a * C::ZETA_SCALAR + b
 }
 
+/// Compute `n(x)` for a given `x`, where `n` is the injective function related to the Halo
```
Comment (dlubarov) [inline src/plonk_util.rs:77]: Seems slightly out of date -- let's say "Compute `[n(s)] P`"

File: src/plonk_util.rs [hunk 4]
```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
...
+/// endomorphism.
+pub(crate) fn halo_n_mul<C: HaloCurve>(s_bits: &[bool], p: AffinePoint<C>) -> AffinePoint<C> {
```
Comment (dlubarov) [inline src/plonk_util.rs:79]: nit: maybe call it halo_endo_mul? Just seems easy to forget what `n` refers to in Halo

File: src/plonk_util.rs [hunk 5]
```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
...
+    let endo_p_p = C::endomorphism(p).to_projective();
+    let endo_p_n = endo_p_p.neg();
+
+    let mut acc = ProjectivePoint::<C>::ZERO;
+
+    for s_bits_chunk in s_bits.chunks(2) {
+        let bit_lo = s_bits_chunk[0];
+        let bit_hi = s_bits_chunk[1];
+
+        let s;
```
Comment (dlubarov) [inline src/plonk_util.rs:95]: nit: maybe slightly nicer as `let s = if bit_hi { ... } else { ... }`, or do you not like that style?

File: src/plonk_util.rs [hunk 6]
```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
...
+        if bit_hi {
```
Comment (dlubarov) [inline src/plonk_util.rs:96]: I think this is switched from the Halo paper, where the first branch is a hi bit of 0. Doesn't really matter but I think I used that convention in the recursive version.

File: src/plonk_util.rs [hunk 7]
```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
...
+                s = p_n;
+            }
+        } else {
+            if bit_lo {
+                s = endo_p_p;
+            } else {
+                s = endo_p_n;
+            }
+        }
+        acc = acc + acc + s;
```
Comment (dlubarov) [inline src/plonk_util.rs:109]: Not very important but `acc.double()` should be a bit faster

File: src/curve.rs [hunk 8]
```diff
@@ -175,19 +205,46 @@ impl<C: Curve> ProjectivePoint<C> {
...
     pub fn add_slices(a: &[Self], b: &[Self]) -> Vec<Self> {
         assert_eq!(a.len(), b.len());
-        a.iter().zip(b.iter()).map(|(&a_i, &b_i)| a_i + b_i).collect()
+        a.iter()
+            .zip(b.iter())
+            .map(|(&a_i, &b_i)| a_i + b_i)
+            .collect()
+    }
+
+    pub fn neg(&self) -> Self {
```
Comment (dlubarov) [inline src/curve.rs:224]: There's also an `impl Neg` for AffinePoint & ProjectivePoint, so `-p` should work

File: src/plonk_util.rs (hunk 4 above)
Comment (dlubarov) [reply src/plonk_util.rs:79]: On second thought, since we do have a `halo_n` method, maybe it's good to match that. Up to you.

Comment (wborgeaud) [reply src/plonk_util.rs:79]: I'll think I'll keep `halo_n_mul` for now to be coherent with the `halo_n` function.

File: src/plonk_util.rs (hunk 6 above)
Comment (wborgeaud) [reply src/plonk_util.rs:96]: Yes you're right. It was switched in the `halo_n` function, so I've changed both `halo_n` and `halo_n_mul` and it should be coherent with the paper and `CurveEndoGate` now.

**Pull request merged**